### PR TITLE
Fix sub-optimal retrieval

### PR DESCRIPTION
**Pull request by nagarev**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

closes https://github.com/rsksmart/rskj/issues/1550

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/trie/Trie.java [hunk 1]
```diff
@@ -766,23 +766,18 @@ private Trie put(TrieKeySlice key, byte[] value, boolean isRecursiveDelete) {
...
+        // both left and right exist (or not) at the same time
+        if (trie.left.isEmpty() == trie.right.isEmpty()) {
             return trie;
         }
 
         Trie child;
         byte childImplicitByte;
-        if (leftOpt.isPresent()) {
-            child = leftOpt.get();
+        if (!trie.left.isEmpty() && trie.left.getNode().isPresent()) {
```
Comment (joaquinlpereyra) [inline rskj-core/src/main/java/co/rsk/trie/Trie.java:776]: My `null` guard comment was regarding `trie.left`, not `trie.left.getNode()`, which will never return `null` as it always returns an `Optional`. This guard calls `getNode()` which is quite expensive unnecessarily (as it has to perform `getHash()`)

This should be taken with a grain of salt though, even though it is possible that `trie.left` is null (nothing in the code prevents it as far as I see), the old code actually does assume that it is not null, maybe it is best to keep it that way.

So:

* To avoid risks of changing behavior I'd keep `trie.left`, `trie.right` without null checks, as the previous code assumed they were always something and this is most likely what's happening if there's no reason to believe otherwise
* The added guard adds more computation and also very subtly modifies behavior.

Comment (nagarev) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java:776]: Previously the code was doing the checkup about the node being not null here:

```java
Optional<Trie> leftOpt = trie.left.getNode();
Optional<Trie> rightOpt = trie.right.getNode();
if (leftOpt.isPresent() && rightOpt.isPresent()) { // <- Checking Here
    return trie;
}
```
so, in the following lines, when we called 

```java
rightOpt.get()
```

We were already covered.

But in the new code we only check that left and right are not empty without checking the contained node _per se_:

```java
if (trie.left.isEmpty() == trie.right.isEmpty()) {
```

So when calling

```java
.getNode().get()
```

we are not sure that the contents are not null.

So I think that:
- one of the points "To avoid risks of changing behavior I'd keep trie.left, trie.right without null checks, as the previous code assumed they were always something..." it's not completely right (because the old code was performing checks on the node being not null).
- we have to decide where and how we want to avoid extra computation. In this case the computation would be made for the cases when we know for sure that one of the left or right references are not empty.

What do you think?

Comment (nagarev) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java:776]: And yes `trie.left` and `trie.right` were not being checked for null right before as you pointed.

Comment (nagarev) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java]: With the checks removed, now if the `Node` element is not present, a `java.util.NoSuchElementException` is thrown by the `get()` method. So I've removed the checks for clarity, and we can start improving the solution from [this new commit](https://github.com/rsksmart/rskj/pull/1575/commits/2580b15ff41217a6abbc50d3bc643f043daae7e3) and on.

Comment (joaquinlpereyra) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java]: Leaving for further discussion: the underlying issue here is that `isEmpty` and `getNode` are not correlated to what one would expect: an `!trie.isEmpty` **does not** assure that `trie.getNode().isPresent()`, at least according to the code.

IMHO this scenario should not be considered an `option`: looking up a node in the db which does not exist should be considered an exception, no correct alternative exists to continue processing the information.

File: rskj-core/src/main/java/co/rsk/trie/Trie.java [hunk 2]
```diff
@@ -766,26 +766,26 @@ private Trie put(TrieKeySlice key, byte[] value, boolean isRecursiveDelete) {
...
+            child = trie.left.getNode().orElse(null);
             childImplicitByte = (byte) 0;
         } else { // has right node
-            child = rightOpt.get();
+            child = trie.right.getNode().orElse(null);
             childImplicitByte = (byte) 1;
         }
 
+        // could not retrieve from database
+        if (child == null) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/trie/Trie.java:785]: why did you add this condition? From my understanding, this should never happen.

Comment (nagarev) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java:785]: Yep, we think the same, but since this "child" is the result of evaluating an `Optional` field, we have to take that in account. Also note that the code behavior before this change was exactly the same, so this is also for the sake of maintaining it the same.

**Comment by nagarev** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Review by Vovchyk — APPROVED** [review]

File: rskj-core/src/main/java/co/rsk/trie/Trie.java (hunk 2 above)
Comment (fedejinich) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java:807]: mmm isn't this an unnecesary check? if you do the `Trie.getNode()` you can get the broken db error

```java
public Optional<Trie> getNode() {
        if (lazyNode != null) {
            return Optional.of(lazyNode);
        }

        if (lazyHash == null) {
            return Optional.empty();
        }

        Optional<Trie> node = store.retrieve(lazyHash.getBytes());

        // Broken database, can't continue
        if (!node.isPresent()) {
            logger.error("Broken database, execution can't continue");
            nodeStopper.stop(1);
            return Optional.empty();
        }

        lazyNode = node.get();

        return node;
    }
```

**Review by fedejinich — COMMENTED** [review]
apart form the comment I've made, I think the overall it's ok
